### PR TITLE
Avoid unnecessary zero-memset of the scratch flags in SYCL

### PR DESCRIPTION
**Pull request by dalg24**
I spotted this oddity in #6732 
SYCL does properly set the flags to zero https://github.com/kokkos/kokkos/blob/34973c77309cb8af20fae065dea7e44146b22aed/core/src/SYCL/Kokkos_SYCL_Instance.cpp#L292-L293

It might very well be that the zero-memset is not needed.

### CODE REVIEW

**Comment by dalg24** [comment]
cc @Rombur @masterleinad @crtrott

**Review by masterleinad — COMMENTED** [review]
This is ultimately internal functionality and so doesn't necessarily need to be consistent across backends (as long as it works correctly with the reduction/scan implementation).
That said whether the function zero-initializes the allocation should be independent of whether we allocate or not.

**Comment by masterleinad** [comment]
For a DOT benchmark, i.e., small functor workload, I' seeing with the current implementation
```
DOT KK 2:       0.0035069 s     0.00084982 GB/s
DOT KK 4:       0.00350057 s    0.00170271 GB/s
DOT KK 8:       0.00372481 s    0.00320041 GB/s
DOT KK 16:      0.00351662 s    0.00677977 GB/s
DOT KK 32:      0.00346489 s    0.013762 GB/s
DOT KK 64:      0.00356406 s    0.0267581 GB/s
DOT KK 128:     0.00350824 s    0.0543676 GB/s
DOT KK 256:     0.00361119 s    0.105636 GB/s
DOT KK 512:     0.104139 s      0.00732616 GB/s
DOT KK 1024:    0.00407179 s    0.374744 GB/s
DOT KK 2048:    0.00443221 s    0.688541 GB/s
DOT KK 4096:    0.00424414 s    1.4381 GB/s
DOT KK 8192:    0.00421679 s    2.89487 GB/s
DOT KK 16384:   0.00424253 s    5.75459 GB/s
DOT KK 32768:   0.00462483 s    10.5578 GB/s
DOT KK 65536:   0.00439912 s    22.199 GB/s
DOT KK 131072:  0.00505756 s    38.6179 GB/s
DOT KK 262144:  0.00506981 s    77.0493 GB/s
DOT KK 524288:  0.0052592 s     148.549 GB/s
DOT KK 1048576: 0.00658116 s    237.42 GB/s
DOT KK 2097152: 0.00688503 s    453.883 GB/s
DOT KK 4194304: 0.00739402 s    845.278 GB/s
DOT KK 8388608: 0.00851514 s    1467.97 GB/s
DOT KK 16777216:        0.0109566 s     2281.72 GB/s
DOT KK 33554432:        0.0335447 s     1490.55 GB/s
DOT KK 67108864:        0.065571 s      1525.06 GB/s
DOT KK 134217728:       0.127886 s      1563.89 GB/s
DOT KK 268435456:       0.263737 s      1516.66 GB/s
DOT KK 536870912:       0.490924 s      1629.58 GB/s
DOT KK 1073741824:      0.95565 s       1674.25 GB/s
```
and when avoiding `memset`:
```
DOT KK 2:       0.00251693 s    0.00118407 GB/s
DOT KK 4:       0.00241424 s    0.00246888 GB/s
DOT KK 8:       0.00249089 s    0.00478581 GB/s
DOT KK 16:      0.00293491 s    0.00812355 GB/s
DOT KK 32:      0.00235021 s    0.0202891 GB/s
DOT KK 64:      0.00238296 s    0.0400206 GB/s
DOT KK 256:     0.00251156 s    0.151886 GB/s
DOT KK 512:     0.00258596 s    0.295031 GB/s
DOT KK 1024:    0.00290403 s    0.525435 GB/s
DOT KK 2048:    0.00295365 s    1.03322 GB/s
DOT KK 4096:    0.00279898 s    2.18062 GB/s
DOT KK 8192:    0.0027936 s     4.36964 GB/s
DOT KK 16384:   0.00290793 s    8.39568 GB/s
DOT KK 32768:   0.00283081 s    17.2488 GB/s
DOT KK 65536:   0.00292013 s    33.4425 GB/s
DOT KK 131072:  0.00352065 s    55.4763 GB/s
DOT KK 262144:  0.00377992 s    103.342 GB/s
DOT KK 524288:  0.00404404 s    193.186 GB/s
DOT KK 1048576: 0.00537093 s    290.918 GB/s
DOT KK 2097152: 0.00555144 s    562.917 GB/s
DOT KK 4194304: 0.0061304 s     1019.51 GB/s
DOT KK 8388608: 0.00717899 s    1741.19 GB/s
DOT KK 16777216:        0.00937119 s    2667.75 GB/s
DOT KK 33554432:        0.0323356 s     1546.28 GB/s
DOT KK 67108864:        0.0645445 s     1549.32 GB/s
DOT KK 134217728:       0.127414 s      1569.68 GB/s
DOT KK 268435456:       0.262706 s      1522.61 GB/s
DOT KK 536870912:       0.490629 s      1630.56 GB/s
DOT KK 1073741824:      0.955163 s      1675.11 GB/s
```
so `memset` takes like 1 ms.
```diff
diff --git a/core/src/SYCL/Kokkos_SYCL_Instance.cpp b/core/src/SYCL/Kokkos_SYCL_Instance.cpp
index 3f931c016..8f4481a5d 100644
--- a/core/src/SYCL/Kokkos_SYCL_Instance.cpp
+++ b/core/src/SYCL/Kokkos_SYCL_Instance.cpp
@@ -298,9 +298,9 @@ sycl::device_ptr<void> SYCLInternal::scratch_flags(const std::size_t size) {
     Record::increment(r);
 
     m_scratchFlags = reinterpret_cast<size_type*>(r->data());
+    auto memset_event = m_queue->memset(m_scratchFlags, 0,
+                                        m_scratchFlagsCount * sizeScratchGrain);
   }
-  auto memset_event = m_queue->memset(m_scratchFlags, 0,
-                                      m_scratchFlagsCount * sizeScratchGrain);
 #ifndef KOKKOS_IMPL_SYCL_USE_IN_ORDER_QUEUES
   m_queue->ext_oneapi_submit_barrier(std::vector{memset_event});
 #endif
diff --git a/core/src/SYCL/Kokkos_SYCL_ParallelReduce_Range.hpp b/core/src/SYCL/Kokkos_SYCL_ParallelReduce_Range.hpp
index 7f29dcf9d..d18528385 100644
--- a/core/src/SYCL/Kokkos_SYCL_ParallelReduce_Range.hpp
+++ b/core/src/SYCL/Kokkos_SYCL_ParallelReduce_Range.hpp
@@ -177,6 +177,8 @@ class Kokkos::Impl::ParallelReduce<CombinedFunctorReducerType,
                 }
                 item.barrier(sycl::access::fence_space::local_space);
                 if (num_teams_done[0] == n_wgroups) {
+		  if (local_id == 0)
+		    *scratch_flags = 0;	  
                   if (local_id >= n_wgroups)
                     reducer.init(&local_mem[local_id * value_count]);
                   else {
```

**Review by crtrott — APPROVED** [review]

**Comment by masterleinad** [comment]
`SYCL` passed and that is the only backend affected by this pull request (apart from comments).

File: core/src/Cuda/Kokkos_Cuda_Instance.cpp [hunk 1]
```diff
@@ -350,6 +350,9 @@ Cuda::size_type *CudaInternal::scratch_flags(const std::size_t size) const {
     m_scratchFlags = static_cast<size_type *>(
         mem_space.allocate("Kokkos::InternalScratchFlags", alloc_size));
 
+    // We only zero-initialiaze the allocation when we actually allocate.
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Instance.cpp:353]:
```suggestion
    // We only zero-initialize the allocation when we actually allocate.
```

File: core/src/HIP/Kokkos_HIP_Instance.cpp [hunk 2]
```diff
@@ -226,6 +226,9 @@ Kokkos::HIP::size_type *HIPInternal::scratch_flags(const std::size_t size) {
     m_scratchFlags = static_cast<size_type *>(
         mem_space.allocate("Kokkos::InternalScratchFlags", alloc_size));
 
+    // We only zero-initialiaze the allocation when we actually allocate.
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Instance.cpp:229]:
```suggestion
    // We only zero-initialize the allocation when we actually allocate.
```

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp [hunk 3]
```diff
@@ -288,12 +288,16 @@ sycl::device_ptr<void> SYCLInternal::scratch_flags(const std::size_t size) {
         m_scratchFlagsCount, sizeScratchGrain);
     m_scratchFlags = static_cast<size_type*>(mem_space.allocate(
         "Kokkos::Experimental::SYCL::InternalScratchFlags", alloc_size));
-  }
-  auto memset_event = m_queue->memset(m_scratchFlags, 0,
-                                      m_scratchFlagsCount * sizeScratchGrain);
+
+    // We only zero-initialiaze the allocation when we actually allocate.
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Instance.cpp:292]:
```suggestion
    // We only zero-initialize the allocation when we actually allocate.
```

**Pull request merged**